python/benchmark/gen_data_distributed.py

```diff
@@ -232,6 +234,155 @@ def make_matrix_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
+        # Output is generated by applying a (potentially biased) random linear regression model to the input, with n_informative
+        # nonzero regressors and some gaussian centered noise with adjustable scale.
+        #
+        # Code adapted from sklearn.datasets.make_regression().
+        if effective_rank is not None:
+            tail_strength = params.get("tail_strength", 0.5)
+            lrm_input_args = [
+                "--num_rows",
+                str(rows),
+                "--num_cols",
+                str(cols),
+                "--dtype",
+                str(dtype),
+                "--output_dir",
+                "temp",
+                "--output_num_files",
+                str(num_partitions),
+                "--effective_rank",
+                str(effective_rank),
+                "--tail_strength",
+                str(tail_strength),
+                "--random_state",
+                str(seed),
+            ]
+            # Generate a low-rank, fat tail input set.
+            X, _ = LowRankMatrixDataGen(lrm_input_args).gen_dataframe(spark)
+            assert X.rdd.getNumPartitions() == num_partitions, (
+                f"Unexpected num partitions received from LowRankMatrix;"
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+        else:
+            # Randomly generate a well-conditioned input set.
+            X = spark.createDataFrame(
+                RandomRDDs.normalVectorRDD(
+                    spark.sparkContext,
+                    rows,
+                    cols,
+                    numPartitions=num_partitions,
+                    seed=seed,
+                ).map(lambda nparray: nparray.tolist()),
+                schema=",".join(self.schema),
+            )
+
+            assert X.rdd.getNumPartitions() == num_partitions, (
+                f"Unexpected num partitions received from RandomRDDs;"
+                f"expected {num_partitions}, got {X.rdd.getNumPartitions()}"
+            )
+
+        # Generate ground truth upfront.
+        ground_truth = np.zeros((cols, n_targets))
+        ground_truth[:n_informative, :] = 100 * generator.uniform(
+            size=(n_informative, n_targets)
+        )
+
+        if shuffle:
+            # Shuffle feature indices upfront.
+            indices = np.arange(cols)
+            generator.shuffle(indices)
+            ground_truth = ground_truth[indices]
+
+        # UDF for distributed generation of X and y.
+        def make_regression_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
+            for pdf in iter:
+                X_p = pdf.to_numpy()
+                n_partition_rows = X_p.shape[0]
+
+                if shuffle:
+                    # Column-wise shuffle (global)
+                    X_p[:, :] = X_p[:, indices]
+
+                y = np.dot(X_p, ground_truth) + bias
+                if noise > 0.0:
+                    y += generator.normal(scale=noise, size=y.shape)
```

Comment:
So due to the unit column normalization of low rank matrix the 'signal to noise' ratio will depend on the number of examples for a given 'noise' parameter.  Is that how sklearn works?

Reply:
Not sure what you mean here by the number of examples for a given noise parameter.

Reply:
Will the variance of the predict value decrease with the number examples generated (e.g. due the orthonormality constraint on the output columns in low-rank-matrix)?

Reply:
Yes, I figure it would. SKLearn's implementation allows this to happen, and they don't ensure any norm on the default cause (with generator.standard_normal) to make the two methods norm-consistent. But in both cases the relative variances of X and y will scale the same, so I imagine it would be fine?

Reply:
And yes, this version is consistent with what sklearn is doing. Normed X with low_rank_matrix, and no norm via the standard normal.